model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java

```diff
@@ -0,0 +1,188 @@
+    private int indexCreationThreshold;
+
+    @Override
+    public SqlStatement[] generateStatements(Database database) {
+        // This check is for manual migration
+        if (ExecutorService.getInstance().getExecutor(database) instanceof LoggingExecutor)
+            return super.generateStatements(database);
+
+        Object indexCreationThreshold = ((AbstractJdbcDatabase) database)
+            .get(DefaultLiquibaseConnectionProvider.INDEX_CREATION_THRESHOLD_PARAM);
+
+        if (indexCreationThreshold instanceof Integer) {
+            this.indexCreationThreshold = (Integer) indexCreationThreshold;
+            if (this.indexCreationThreshold <= 0)
+                return super.generateStatements(database);
+        } else {
+            return super.generateStatements(database);
+        }
+        try {
+            if (!SnapshotGeneratorFactory.getInstance()
+                .has(new Table().setName(getTableName()).setSchema(new Schema(getCatalogName(), getSchemaName())), database))
+                return super.generateStatements(database);
+
+            int result = ExecutorService.getInstance().getExecutor(database).queryForInt(
+                new RawSqlStatement("SELECT COUNT(*) FROM " + getTableNameForSqlSelects(database, getTableName())));
+
+            if (result > this.indexCreationThreshold) {
+                String loggingString = createLoggingString(database);
+                logger.warnv("Following index should be created: {0}", loggingString);
+                getChangeSet().setComments(loggingString);
+                return new SqlStatement[] {};
+            }
+
+        } catch (DatabaseException | InvalidExampleException e) {
+            throw new UnexpectedLiquibaseException("Database error while index threshold validation.", e);
+        }
+
+        return super.generateStatements(database);
+    }
+
+    private String getTableNameForSqlSelects(Database database, String tableName) {
+        String correctedSchemaName = database.escapeObjectName(database.getDefaultSchemaName(), Schema.class);
+        return LiquibaseJpaUpdaterProvider.getTable(tableName, correctedSchemaName);
+    }
+
+    private String createLoggingString(Database database) throws DatabaseException {
+        StringWriter writer = new StringWriter();
+        LoggingExecutor loggingExecutor = new LoggingExecutor(ExecutorService.getInstance().getExecutor(database), writer,
+            database);
+        SqlStatement sqlStatement = new CreateIndexStatement(getIndexName(), getCatalogName(), getSchemaName(), getTableName(),
+            this.isUnique(), getAssociatedWith(), getColumns().toArray(new AddColumnConfig[getColumns().size()]))
+                .setTablespace(getTablespace()).setClustered(getClustered());
+
+        loggingExecutor.execute(sqlStatement);
+
+        return writer.toString();
+    }
+
+    @Override
+    public boolean generateStatementsVolatile(Database database) {
```

Comment:
Can we avoid overriding this and following two methods?

Reply:
Without these 3 override methods and without expressly calling the `generateStatements` method of the superclass, the `generateStatements` method of this class is called 14 times in one migration (so the warning outputs 14 times).
I also considered the way to add if which judges the caller method by using stack trace in the `generateStatements` method, but I think the current way is smarter.
If there is a better way, I'm happy if you tell me.